monitor/ethereum.go

```
@@ -544,9 +626,19 @@ func (m *Monitor) getEthereumBalance(apiUrl, address, apiKey, contractAddress st
 		params.Add("action", "balance")
 	}
 
+	// New Etherscan API v2 requires chainId to be specified in the URL
+	// if the apiUrl has /v2/ in the path, add chainId to the URL
+	if strings.Contains(apiUrl, "v2") {
+		params.Add("chainid", strconv.Itoa(chainId))
+	}
+
 	url := fmt.Sprintf("%s?%s", apiUrl, params.Encode())
 	req, err := http.NewRequest("GET", url, nil)
 	if err != nil {
+		if req.Response.StatusCode == http.StatusTooManyRequests {
```

Why do you need to retry, the `RunAll` will attempt to run every minute. It's configured using a ticker in `main.go`.

```go
	ticker := time.NewTicker(time.Duration(*interval) * time.Minute)
	defer ticker.Stop()

	tickerHourly := time.NewTicker(time.Hour)
	defer tickerHourly.Stop()

	sigs := make(chan os.Signal, 1)
	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
```

Just let it fail, you'll pick it up on the next interval. You can increase the interval duration if you are facing issues - like doing it every 10-15 minutes instead of every minute.

No retry mechanism is necessary for something like this.

The one second backoff is probably too low anyway because the lowest backoff time on etherscan products is a minute or something like that.

---

i was hitting condition where data was never collected even on retry because it was rate limiting from balances or txs calls...and since there's no logic to deal with it the data in the DB was very old
I dont get why you wouldnt want retries or some kind of conditional handling?
I get this is far from perfect but there are conditions where data is never collected because you hope it happens on the next run